src/main/utility/childpid_watcher.rs

```diff
@@ -0,0 +1,465 @@
+        let command_notifier =
+            nix::sys::eventfd::eventfd(0, nix::sys::eventfd::EfdFlags::EFD_NONBLOCK).unwrap();
+        let mut event = EpollEvent::new(EpollFlags::EPOLLIN, 0);
+        epoll_ctl(
+            epoll,
+            EpollOp::EpollCtlAdd,
+            command_notifier,
+            Some(&mut event),
+        )
+        .unwrap();
+        let watcher = ChildPidWatcher {
+            inner: Arc::new(Mutex::new(Inner {
+                next_handle: 1,
+                pidfds: HashMap::new(),
+                callbacks: HashMap::new(),
+                commands: Vec::new(),
+                command_notifier,
+                thread_handle: None,
+            })),
+            epoll,
+        };
+        let thread_handle = {
+            let inner = watcher.inner.clone();
+            let epoll = watcher.epoll;
+            thread::Builder::new()
+                .name("child-pid-watcher".into())
+                .spawn(move || ChildPidWatcher::thread_loop(&*inner, epoll))
+                .unwrap()
+        };
+        watcher.inner.lock().unwrap().thread_handle = Some(thread_handle);
+        watcher
+    }
+
+    fn thread_loop(inner: &Mutex<Inner>, epoll: RawFd) {
+        let mut events = [EpollEvent::empty(); 10];
+        let mut commands = Vec::new();
+        let mut done = false;
+        while !done {
+            let nevents = epoll_wait(epoll, &mut events, -1).unwrap();
+
+            // We hold the lock the whole time we're processing events. While it'd
+            // be nice to avoid holding it while executing callbacks (and therefor
+            // not require that callbacks don't call ChildPidWatcher APIs), that'd
+            // make it difficult to guarantee a callback *won't* be run if the
+            // caller unregisters it.
+            let mut inner = inner.lock().unwrap();
+
+            for event in &events[0..nevents] {
+                let pid = Pid::from_raw(i32::try_from(event.data()).unwrap());
+                inner.run_callbacks_for_fd(epoll, pid);
+            }
+            // Reading an eventfd always returns an 8 byte integer. Do so to ensure it's
+            // no longer marked 'readable'.
+            let mut buf = [0; 8];
+            let res = nix::unistd::read(inner.command_notifier, &mut buf);
+            debug_assert!(match res {
+                Ok(8) => true,
+                Ok(i) => panic!("Unexpected read size {}", i),
+                Err(nix::Error::Sys(nix::errno::Errno::EAGAIN)) => true,
+                Err(e) => panic!("Unexpected error {:?}", e),
+            });
+            // Run  commands
+            std::mem::swap(&mut commands, &mut inner.commands);
+            for cmd in commands.drain(..) {
+                match cmd {
+                    Command::RunCallbacks(pid) => {
+                        inner.run_callbacks_for_fd(epoll, pid);
+                    }
+                    Command::Finish => done = true,
```

Comment:
After receiving a `Finish` command, it runs the callbacks for the remaining commands, but not the callbacks for remaining ready epoll events. I think it would be better to be consistent here and either return immediately (skipping all remaining callbacks), or process all remaining callbacks from commands *and* ready epoll events before returning.

Reply:
We already processed epoll events above, but I suppose there could always be more (either because there were more pending than the size of buffer we provided or more arrived in the meantime).

I think just exiting here is fine though, and perhaps a bit clearer. Done.